File: Part 2 - Regression/Section 7 - Support Vector Regression (SVR)/svr.R
# Importing the dataset
dataset = read.csv("Position_Salaries.csv")
dataset = dataset[2:3]

# Splitting the dataset in train set and test set
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Profit, SplitRatio = 0.8)
# train_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature scaling
# a biblioteca aplica feature scalling internamente, assim como no python
# train_set$Country = as.numeric(train_set$Country)
# test_set$Country = as.numeric(test_set$Country)
# train_set[,1:3] = scale(train_set[,1:3])
# test_set[,1:3]  = scale(test_set[,1:3])

# SVR

# Fitting SVR to the dataset
library(e1071)
reg = svm(formula = Salary ~ .,
          data = dataset,
          type = "eps-regression")
summary(reg)

# Prediction a new reuslt with SVR
x = data.frame(Level = 6.5)
Y_pred = predict(reg, newdata = x)

# Visualising the SVR results 
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +
  geom_line(aes(x = dataset$Level, y = predict(reg, newdata = dataset)),
            colour = "blue") +
  geom_point(aes(x = 6.5, y = predict(reg, newdata = data.frame(Level = 6.5))),
             colour = "green") +
  ggtitle("SVR") +
  ylab("Level") +
  xlab("Salary")

# Visualising the SVR results (grid)
grid = seq(1, 10, by=0.1)
X_grid = data.frame("Level" = grid)

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +
  geom_line(aes(x = X_grid$Level, y = predict(reg, newdata = X_grid)),
            colour = "blue") +
  geom_point(aes(x = 6.5, y = predict(reg, newdata = data.frame(Level = 6.5))),
             colour = "green") +
  ggtitle("Polynomial SVR (grid = 0.1)") +
  ylab("Level") +
  xlab("Salary")

File: Part 2 - Regression/Section 7 - Support Vector Regression (SVR)/svr.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv("Position_Salaries.csv")
X = dataset.iloc[:, 1:-1].values
#coloquei : no y para virar uma matrix de uma coluna, pois é requisito para o feature scalling
Y = dataset.iloc[:, -1:].values

# Splitting the dataset in train set and test set
"""
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)
"""

# Feature scaling
# a biblioteca NÂO aplica feature scalling internamente automaticamente
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_Y = StandardScaler()
X = sc_X.fit_transform(X)
Y = sc_Y.fit_transform(Y)

# SVR

# Fitting SVR to the dataset
from sklearn.svm import SVR
reg = SVR(kernel = "rbf")
reg.fit(X,Y)

# Prediction a new result with SVR
# a escala do problema mudou por causa do feature scaling, logo a entrada x deve ser sofre o mesmo processo
x = sc_X.transform(np.array(6.5).reshape(-1,1))
Y_pred = reg.predict(x)
# a previsão é feita no novo sistema de escala, mas queremos vê-la em sua escala original
Y_pred = sc_Y.inverse_transform(Y_pred)

# Visualising the Regression results
plt.scatter(X,Y, color = "red")
plt.plot(X, reg.predict(X), color = "blue")
plt.scatter(x, reg.predict(x), color = "green")
plt.title("SVR")
plt.xlabel("Position level")
plt.ylabel("Salary")
plt.show()

# Visualising the SVR results (grid)
X_grid = np.arange(min(X),max(X), step = 0.1)# vetor
X_grid = X_grid.reshape((len(X_grid),1))# matriz

plt.scatter(X,Y, color = "red")
plt.plot(X_grid, reg.predict(X_grid), color = "blue")
plt.scatter(x, reg.predict(x), color = "green")
plt.title("SVR (grid = 0.1)")
plt.xlabel("Position level")
plt.ylabel("Salary")
plt.show()

